modules/generators.py

```diff
@@ -154,10 +158,14 @@ async def _gpt_basic_factory(
         )
         good_generations += [
             g
-            for g in generations.outputs
+            for g in generations.outputs or []
             if g.function_call is None or g.function_call["name"] in agent.toolkit_dict
         ]
         middleman_settings_copy.n = num_to_generate - len(good_generations)
+
+    if generations is None:
+        raise ValueError("No generations returned from gpt_basic_factory")
```

Comment:
Is this OK?

Reply:
IIUC this would never trigger because you can only break out of the loop above if generations is not None. In either case if you've successfully broken out of the loop, `good_generations` is what you really care about, so I think we can remove this.

Reply:
`generations` is called below on line 180: `generations.dict().items()`